File: frontend/src/api/tutoring.ts
import request from './request'
import { getToken } from '@/utils/storage'
import { resolveApiBaseURL } from '@/api/baseUrl'

export type TutoringMode = 'ask' | 'hint' | 'explain'
export type TutoringDifficulty = 'easy' | 'normal' | 'advanced'
export type TutoringResponseFormat = 'markdown' | 'plain'

export interface TutoringKnowledgeOptions {
  course_id?: number | null
  knowledge_document_ids?: number[] | null
  use_knowledge_base?: boolean
  top_k?: number
  difficulty?: TutoringDifficulty
}

export interface TutoringAskRequest extends TutoringKnowledgeOptions {
  question: string
  response_format?: TutoringResponseFormat
}

export interface TutoringHintRequest extends TutoringKnowledgeOptions {
  question: string
  context?: string | null
  response_format?: TutoringResponseFormat
}

export interface TutoringExplainRequest extends TutoringKnowledgeOptions {
  concept: string
  response_format?: TutoringResponseFormat
}

export interface TutoringReference {
  document_id?: number | null
  chunk_index?: number | null
  source_filename?: string | null
  excerpt?: string | null
  score?: number | null
}

export interface TutoringResponse {
  session_id: number
  question?: string
  concept?: string
  answer?: string
  hint?: string
  explanation?: string
  references: TutoringReference[]
  warnings?: string[]
  used_knowledge_base: boolean
  response_format: TutoringResponseFormat
  created_at: string
}

export interface TutoringSession {
  id: number
  user_id: number
  course_id?: number | null
  topic?: string | null
  session_type: TutoringMode | string
  user_question: string
  ai_response: string
  response_format: string
  context_refs: TutoringReference[]
  is_helpful?: boolean | null
  user_rating?: number | null
  created_at: string
  updated_at: string
}

export interface TutoringSessionListResponse {
  items: TutoringSession[]
  total: number
  page: number
  page_size: number
}

export interface TutoringRatingRequest {
  is_helpful: boolean
  user_rating: number
}

export interface TutoringRatingResponse {
  session_id: number
  is_helpful: boolean
  user_rating: number
}

export function askTutoring(payload: TutoringAskRequest) {
  return request.post<TutoringResponse, TutoringResponse>('/student/tutoring/ask', payload)
}

export function getTutoringHint(payload: TutoringHintRequest) {
  return request.post<TutoringResponse, TutoringResponse>('/student/tutoring/hint', payload)
}

export function explainConcept(payload: TutoringExplainRequest) {
  return request.post<TutoringResponse, TutoringResponse>('/student/tutoring/explain', payload)
}

export function getTutoringSessions(params?: {
  page?: number
  page_size?: number
  topic?: string
  session_type?: TutoringMode | ''
}) {
  return request.get<TutoringSessionListResponse, TutoringSessionListResponse>('/student/tutoring/sessions', { params })
}

export function listTutoringSessions(params?: Record<string, unknown>) {
  return request.get('/student/tutoring/sessions', { params })
}

export function rateTutoringSession(sessionId: number, payload: TutoringRatingRequest) {
  return request.post<TutoringRatingResponse, TutoringRatingResponse>(`/student/tutoring/sessions/${sessionId}/rating`, payload)
}

export interface TutoringMessageRecord { id:number; conversation_id:number; role:'user'|'assistant'; content:string; status:string; references:TutoringReference[]; warnings:string[]; error?:string|null; client_message_id?:string|null; created_at:string; updated_at:string }
export interface TutoringConversation { id:number; user_id:number; course_id?:number|null; title:string; messages:TutoringMessageRecord[]; created_at:string; updated_at:string }
export type TutoringStreamEvent = { type:'meta'|'delta'|'reference'|'warning'|'done'|'error'; text?:string; error?:string; assistant_message_id?:number; message?:TutoringMessageRecord }

export function createTutoringConversation(payload: { title:string; course_id?:number|null }) { return request.post<TutoringConversation,TutoringConversation>('/student/tutoring/conversations',payload) }
export function getTutoringConversations() { return request.get<TutoringConversation[],TutoringConversation[]>('/student/tutoring/conversations') }
export function getTutoringConversation(id:number) { return request.get<TutoringConversation,TutoringConversation>(`/student/tutoring/conversations/${id}`) }
export async function streamTutoringMessage(id:number,payload:{content:string;client_message_id:string;retry_assistant_message_id?:number|null},signal:AbortSignal) {
  const baseURL=resolveApiBaseURL(); const token=getToken()
  const response=await fetch(`${baseURL}/student/tutoring/conversations/${id}/messages/stream`,{method:'POST',headers:{'Content-Type':'application/json',Accept:'application/x-ndjson',...(token?{Authorization:`Bearer ${token}`}:{})},body:JSON.stringify(payload),signal})
  if(!response.ok||!response.body) throw new Error(`辅导流连接失败：HTTP ${response.status}`)
  return response
}

File: backend/app/schemas/tutoring.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


ResponseFormat = Literal["markdown", "plain"]
TutorDifficulty = Literal["easy", "normal", "advanced"]
TutoringSessionType = Literal["ask", "hint", "explain"]


class TutoringKnowledgeMixin(BaseModel):
    course_id: int | None = Field(default=None, description="Optional course filter")
    knowledge_document_ids: list[int] | None = Field(default=None, description="Optional ingested knowledge document ids")
    use_knowledge_base: bool = Field(default=True, description="Whether to retrieve Chroma context")
    top_k: int = Field(default=5, ge=1, le=10, description="Number of chunks to retrieve")
    difficulty: TutorDifficulty = Field(default="normal", description="Tutoring explanation level")


class TutoringAskRequest(TutoringKnowledgeMixin):
    question: str = Field(min_length=2, max_length=1000, description="Student question")
    response_format: ResponseFormat = Field(default="markdown", description="Answer format")


class TutoringHintRequest(TutoringKnowledgeMixin):
    question: str = Field(min_length=2, max_length=1000, description="Student question")
    context: str | None = Field(default=None, max_length=3000, description="Student's current context or confusion")
    response_format: ResponseFormat = Field(default="markdown", description="Hint format")


class TutoringExplainRequest(TutoringKnowledgeMixin):
    concept: str = Field(min_length=1, max_length=255, description="Concept to explain")
    response_format: ResponseFormat = Field(default="markdown", description="Explanation format")


class TutoringReference(BaseModel):
    document_id: int | None = None
    chunk_index: int | None = None
    source_filename: str | None = None
    excerpt: str | None = None
    score: float | None = None


class TutoringAskResponse(BaseModel):
    session_id: int
    question: str
    answer: str
    references: list[TutoringReference] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    used_knowledge_base: bool
    response_format: ResponseFormat
    created_at: datetime


class TutoringHintResponse(BaseModel):
    session_id: int
    question: str
    hint: str
    references: list[TutoringReference] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    used_knowledge_base: bool
    response_format: ResponseFormat
    created_at: datetime


class TutoringExplainResponse(BaseModel):
    session_id: int
    concept: str
    explanation: str
    references: list[TutoringReference] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    used_knowledge_base: bool
    response_format: ResponseFormat
    created_at: datetime


class TutoringSessionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    course_id: int | None = None
    topic: str | None = None
    session_type: str
    user_question: str
    ai_response: str
    response_format: str
    context_refs: list
    is_helpful: bool | None = None
    user_rating: float | None = None
    created_at: datetime
    updated_at: datetime


class TutoringSessionListResponse(BaseModel):
    items: list[TutoringSessionRead]
    total: int
    page: int
    page_size: int


class TutoringRatingRequest(BaseModel):
    is_helpful: bool
    user_rating: int = Field(ge=1, le=5, description="Rating from 1 to 5")


class TutoringRatingResponse(BaseModel):
    session_id: int
    is_helpful: bool
    user_rating: float


class TutoringConversationCreate(BaseModel):
    course_id: int | None = None
    title: str = Field(default="新辅导会话", min_length=1, max_length=255)


class TutoringMessageCreate(BaseModel):
    content: str = Field(min_length=1, max_length=5000)
    client_message_id: str = Field(min_length=1, max_length=100)
    retry_assistant_message_id: int | None = None
    use_knowledge_base: bool = True
    top_k: int = Field(default=5, ge=1, le=10)


class TutoringMessageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int; conversation_id: int; role: str; content: str; status: str
    references: list[dict[str, Any]] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    error: str | None = None
    client_message_id: str | None = None
    created_at: datetime; updated_at: datetime


class TutoringConversationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int; user_id: int; course_id: int | None = None; title: str
    messages: list[TutoringMessageRead] = Field(default_factory=list)
    created_at: datetime; updated_at: datetime
